### ISSUE TITLE
Roadmap brainstorm (5.8+)

### ISSUE THREAD

**Issue by Zughy** [minetest/minetest#12746]
It's been ca. 2 years since the first roadmap [came into play](https://github.com/minetest/minetest/pull/10536). Considering that priorities could be different compared to two years ago, **please list 3 things that you think should be prioritised in the medium-long term**.

These are typically higher-level goals rather than individual features. The new roadmap will come into play with MT 5.8.

**Comment by Zughy** [comment]
My two cents:

### UI Improvement
Although PRs like #12480 are heading in the right direction, I think a lot of work shall be done still when it comes down to chat and main menu (see #10372 for the former and #6733 for the latter). A bad UI/UX means a bad first impression, which means that people will be skeptical and more hard to convince

### Different dimensions (#4428)
I can't stress enough how much of a game-changer this is. Right now people can't instantiate disposable worlds (e.g. dungeons, mini-games, events) and, for anyone who wants to create the illusion of different dimensions, they must stick them one on top of the other. Which I find very unpractical

### Prioritise bugfixes
Right now there are [305 confirmed bugs](https://github.com/minetest/minetest/issues?page=1&q=is%3Aissue+is%3Aopen+label%3ABug). I know it's more thrilling to implement a new feature rather than to fix a bug, but I think that a clean desk makes working easier than a messy one. In these last months we've started gaining control of PRs (we used to have 150+ PRs, now we have 80) and issues (1100+, now less than 1000): if we can gain control of bugs too, I'm prone to think implementing new features will be a lot easier than it is now. And people will be in general less stressed, which is always a good thing :heart: In short: crap first, then fun

**Comment by oong819** [comment]
The Minetest Game will have amother dimension?
or just multi dimension support?

**Comment by IrrlichtMt4** [comment]
Some things from me are:
1. Add a way to modify how objects are shaped using Lua as well/model using Lua. 
2. I think modders should be able to control physics completely.

**Comment by appgurueu** [comment]
# Entity Performance Improvements

Entities are required for most interesting things in the game. Unfortunately, Minetest can only handle very few entities very well. Networking should be optimized to only send changes (for properties for example); the API should be optimized to allow getting/setting single properties efficiently. Spatial indexing should be used to speed up range- and raycast queries. Rendering should batch entities of the same visuals to reduce drawcalls.

# [Network Encryption](https://github.com/minetest/minetest/issues/10206)

The conflict with performance is not really there, and encryption is a basic thing for (1) privacy and (2) security as admins might want to interface with the server without risking others hijacking their sessions.

# (SS?)CSM

This carries over from the old roadmap. Modders need to be able to control client code to some extent. CSM capabilities must be vastly extended for this to be useful. A low-level 2d & 3d rendering access would be great; only then will we slowly approach being a game engine.

**Comment by IrrlichtMt4** [comment]
Hi, may I ask why my comment was marked as off-topic?

**Comment by Emojigit** [comment]
> Hi, may I ask why my comment was marked as off-topic?

“Please list 3 things that you think should be prioritized in the medium-long term.”

**Comment by cranezhou** [comment]
minetest game engine could be used for Education. 
What is the latest status of framinetest ? (https://framinetest.org/en/)
Try to get the feedback from the teachers and the junior, senior and university students groups.
Then we may be able to find the users to drive the development team to move on their roadmap(s). 

Good design, technical enhancement and user(ui) experience are very important.
Data Privacy (GDPR) and Users Login Agreement are also very important.
Is there a gap between the Commercial Products and Open Source Products ?

**Comment by LuciusofLegend** [comment]
### Modding Guide
Modding accessibility is a priority of mine at the moment (especially since I'm a new modder myself ;) ).  Here are some ideas on how to improve it:

- Port the Modding Book to MkDocs with a nice theme (like the Lua API Docs) for easier reading
- Explain essential best practices for compatibility and optimization
- Improve organization from the Modding Book (no offense rubenwardy, improvement can always be made :1st_place_medal: )
### Assent
I agree with everything mentioned before, prominently including Zughy's bit about UI improvements, and appgurueu's about network encryption.

**Comment by LuciusofLegend** [comment]
What about my comment is confusing?  I'll delete this message either when I get an answer or in 2 days.  Sound reasonable?

**Comment by Wuzzy2** [comment]
1. The focus should be on fixing bugs, preferably ancient bugs that are many years old. It WILL pay off in the long run. This is easier said then done so additionally I propose that a proper priotization system is established. Most bugs have NO priority assigned to them.
Each bug could roughly be rated by severity, it is a rating that roughly describes how painful this bug is to players, users, modders, devs, etc. The severity levels must be defined by clearly determinable and reproducible criteria so it can be quickly assigned by reading the text alone (in most cases) and to avoid endless arguments whether a bug is of high or medium severity. The current label system of "high priority", "low priority" and "(no label)" is too vague and needs change. I propose the levels "critical", "very high", "high", "medium" and "low". Why two high priority levels? Because "high priority" is currently used for only the SUPER important things. Therefore, all the important but not SUPER important things drown in the ocean of unlabelled issued. The label for medium is added because this weird implication of "no label = medium" is very annoying.
Another method to get the bug count down is to require a minimum of information of the reporter. If the bug reported is too vague, poorly written and the reporter does not reply after X days, just close the bug. It makes no sense to keep bad bug reports open forever because they MIGHT be true. Just close them, otherwise they just clog up the system

2. Define backwards-compability rules and take them seriously. In my opinion, I would like to see more strictness. Avoid breakage, even small ones, like the plague, EXCEPT for major releases in which it is allowed. A game written for 5.6.0 should still work perfectly in 5.10.0 without ANY additional work from the developer required. Minetest was pretty compatible in the past versions, but it was not perfect, so I like to see affirmation for strong backwards-compability. Also, breakage should really only be considered when all other possibilities have been exhausted. PRs should also be rejected on the basis of them breaking stuff for no good reason. Before a non-major release, all bugs labelled "Regression" must be fixed (if this rule is not already in place).

3. Main menu rework, both in terms of looks/graphics as well in usability because it is a visual abomination that looks like a tech demo

**Comment by rubenwardy** [comment]
Just want to remind people that bugs are not subject to the roadmap, bug fixes are already priority and any related suggestions should be discussed elsewhere

**Comment by appgurueu** [comment]
@Wuzzy2 I believe the purpose of this issue is to determine which feature requests to prioritize for 5.8.

**Comment by PrairieAstronomer** [comment]
> I believe the purpose of this issue is to determine which feature requests to prioritize for 5.8.

Err, Zughy also mentioned prioritizing bug fixes, but now that I read rubenwardy's post, fair enough complaining.

**Comment by cranezhou** [comment]
Sorry, my comment may be a little high level which caused the confusion.
By the way, personally I think the minecraft may also have a lot of lessons, which minetest could learn from.
Thank you very much for your understanding and support. 🙂

**Comment by oong819** [comment]
**Optimize RAM Use**
majority of Minetest's games crash when i load a world with it because of RAM overflow, the same thing happen on some big mod and too many mods installed, Minetest with me are not devices friendly, i have 1,5 GB RAM and can play similar game normally

**More Beginner Friendly**
when i start playing Minetest, i dont evem know how to fly and grant privs, i think Minetest should have a tutorial world for hard things like:
Commands
Controls
PICK UP CART ON MOBILE
Introducing Minetest Game and Content DB
..v..v..
Beginner is likely to only play MTG and think Minetest is a game not game engine

**Children Friendly Features**
to keep our babys brain clean, add things like:
Editable client side chat filter
Add "Family Friendly" certificate for server that are family friendly (probally need a strict rules for this)

**Comment by appgurueu** [comment]
> majority of Minetest's games crash when i load a world with it because of RAM overflow, the same thing happen on some big mod and too many mods installed, Minetest with me are not devices friendly, i have 1,5 GB RAM and can play similar game normally
This sounds like (1) memory leaks in mods and/or (2) the LuaJIT 1GB memory limitation, both of which aren't really issues Minetest can fix (well, the latter can be fixed by using PUC Lua 5.1 or applying a certain patch to LuaJIT, both at the cost of performance)...

**Comment by oong819** [comment]
> > majority of Minetest's games crash when i load a world with it because of RAM overflow, the same thing happen on some big mod and too many mods installed, Minetest with me are not devices friendly, i have 1,5 GB RAM and can play similar game normally
> 
> This sounds like (1) memory leaks in mods and/or (2) the LuaJIT 1GB memory limitation, both of which aren't really issues Minetest can fix (well, the latter can be fixed by using PUC Lua 5.1 or applying a certain patch to LuaJIT, both at the cost of performance)...

i rather live with limited mods than device got piping hot

**Comment by AFCMS** [comment]
### SSCSM

We really need this. Both as modder and game developer.

- Animated HUDs (without fulfilling the server's bandwidth with HUD_CHANGE packets)  
- Client generated formspecs (fixing pandorabox inventory taking 5 seconds to update the crafting guide search)
- More fluid controls (W-tap sprinting, like in the mcl2 CSM)

Would improve server's performances and global user experience.

Should be **really** prioritized IMO.

### CSM

CSM have only few use-cases at the moment.

You can definitely make great things, but it is still much less powerful than it should.

CSM would need exactly the same permissions SSM currently have on the host machine: access to insecure env, http API, filesystem, etc

What could be done by just giving CSM as many permissions as SSM:
- Discord RPC
- RGB Driver
- etc

#12398 

### Main Menu

Current main menu is really really ugly. Recently, a big (+1M) french youtuber made a video about minetest, and his first reaction when starting the game was "Its ugly".

Main menu have a BIG impact on how users see Minetest, its in fact the part of minetest you see the most often in your minetest journey.

It is kinda weird to see how beautiful the i3 mod is compared to the actual main menu...

**Comment by sfan5** [comment]
#### Dehardcoding 

This describes taking an - often client-side graphical - feature that is fixed into the code and adding an API for it so it can be controlled from Lua, a simple example is #2139. Others would be the upcoming camera API, wieldhand animations, object death response and lots of subtle things that are currently client-side settings.
For this to happen efficiently the affected features should be collected in a list first, reading the source here can help spot a lot of harcoded things.

#### Performance and general improvements

This applies especially to graphics where often a "simple" shader-first rewrite could raise the performance from 'terribly bad' to 'blazing fast' (e.g. particles) but also applies to other parts of the code (e.g. SAO collision detection).
This section can also include "changes that resolve a problem that is not necessarily a bug, without creating new real problems by design" as @Desour put it. Though I would count code architecture improvements under here too.

#### &lt;free spot&gt;

dunno

**Comment by Wuzzy2** [comment]
I like dehardcoding a lot, the more is left up to mods, the better. Maybe even add a new label to mark issues/PRs about dehardcoding. I could help with adding the label to issues (once the label exists, I don't think I'm allowed to create labels).

And there is already a list of hardcoded features, you can find it at <https://dev.minetest.net/List_of_hardcoded_features>.

**Comment by TurkeyMcMac** [comment]
I don't have much of a roadmap myself, but I think increasing performance is a good goal, especially as someone without a powerful computer. For my part, there are several places in the backend where easy performance gains are possible. There is also rendering etc. which I haven't worked on but which I hear is ripe for performance improvement.

**Comment by fluxionary** [comment]
i'm going to put a single suggestion here, as a request from a large number of players on the your-land server.

improve support for touch-screen (mobile) users. 

computers w/ keyboards are becoming less common, and more and more people can only play the game on their phones or tablets. if we want the playerbase for this game to continue to grow, this is something that we cannot afford to ignore. i (unfortunately?) don't own such a device that can play this game, so i have to speak from anecdotes, but it seems that the touch/mobile experience is lacking in a lot of respects. 

players complain that GUI scaling is terrible. scrolling various things is terrible. using modifier keys is terrible. here's some existing bug reports that i've just dug up, w/out fully understanding them:
* [Android: Can't use sneak/aux while digging or punching](https://github.com/minetest/minetest/issues/11404)
* [Allow switching between touchscreen and other input on Desktop and Android](https://github.com/minetest/minetest/issues/10062)
* [Add API to get user input mechanisms (ie: touchscreen? gamepad? keyboard?)](https://github.com/minetest/minetest/issues/12264)
* [Android: bad GUI scalling / poor use of available touch screen](https://github.com/minetest/minetest/issues/12227)
* this is just from one page, there are many more

as a mod writer, i wish i had a mechanism to even detect a player's display size and input capabilities, and react to that, but instead i just have to use formspec and pray that it'll be visible. 

as i dislike writing any UI that isn't a programmatic API, this is not an issue i want to deal w/ myself. however, i feel it is incredibly important for the future of the game.

**Comment by velartrill** [comment]
i'd personally like to see SSCSM entirely **replace** formspecs, and for the latter to be deprecated (possibly reimplemented as a wrapper around the notional new lua UI kit to avoid maintaining multiple GUI subsystems indefinitely). i recall seeing discussion/preliminary work being done on this a year or so ago but i don't know what became of that. formspec programming is just miserable, among the worst experiences i've ever _had_ building interactive GUIs. storing cached formspecs as strings in node metadata is a particularly unpleasant pattern i would like to see the end of. (ive designed & implemented a handful of UI toolkits before so i have some suggestions on this topic, and would be happy to do some of the design & coding if others agree with me on this.)

but, i'd be happy just to see some level of SSCSM in 5.8. my UI suggestion might end up being a better fit for a later version, depending

**Comment by Wuzzy2** [comment]
As much as burning formspecs to the ground is a noble goal, this is not that simple. Tons of games and mods depend on this, so a complete removal means breaking almost all games and mods ever made. Nothing short of a hard fork will get you there. A full removal of formspecs is not going to happen.

**Comment by velartrill** [comment]
@Wuzzy2 which is why i never suggested such a thing. "deprecate" does not mean "excise"

**Comment by nerzhul** [comment]
My 50 cents, i put it in touch like every roadmap
* better enties, with more intelligent API to transfer actions from server to client. We should not have for example to run per movement code server side and set each movement to client, a jump should be a jump, a move from A to B if using the path finder should use path finder on client. Obviously with entity pos resync every 5 sec or something. This world is so boring because we don't have more high level API :) look at all those minecraft things (breaking, obviously, if a server uses it)
* Multiple maps, with different mapgen for example (breaking, obvisouly, if server has multiple layers
* Refactoring, clearly, those Server, (Local)Player, Game and Client classes are always too heavy, we should find better objects
* Performance, i agree, it should come from refactoring too. Maybe we can have measures on the code. C++ is less natively tooled than rust, golang and java unfortunately
* WASM moding ? (it's more a tought, not sure it's possible easily for our game with good security)

**Comment by CowboyLva** [comment]
Few points/categories come to mind.

1. Controller support
2. Better map related stuff - Optimized mapgen, definable world size, larger possible worlds at least xz axes, per map configuration menu(e.g. how far each axis will generate), easily customizable generation distance(not render distance).
3. Mod branching in ContentDB(either one creator has many versions of their mod, or many people have fork of something), although just maybe since could be bloat.
4. Main menu changed, but kept simple and also a chance to revert to old or oldish layout.
5. More shadow angle customization(as of now sun is only Northwards, even at 0 the shade side is South).

**Comment by appgurueu** [comment]
> 3. Mod branching in ContentDB(either one creator has many versions of their mod, or many people have fork of something), although just maybe since could be bloat.

CDB already hosts multiple versions of the same mod (although this is usually hidden and you can only install/upgrade to the latest version). Forks are allowed as well.

> 4. Main menu changed, but kept simple and also a chance to revert to old or oldish layout.

Ideally I'd hope for the main menu to be customizable comparable to mods; perhaps main menus could even be distributed through CDB? The `mainmenu_script` setting only works partially.

> 5. More shadow angle customization(as of now sun is only Northwards, even at 0 the shade side is South).

Sounds like a bug to me?

**Comment by velartrill** [comment]
@appgurueu gonna channel hecktest a bit here and suggest that any main menu customization should really be up to the game that's selected, not just random mods people can download. tho i think a case could be made for CSMs that add new tabs -- they just shouldn't have the power to mess with game setup interfaces. there should probably also be guidelines to maximize consistency between different main menu designs, and the layout should make clear which part of the screen belongs to the engine vs. the selected game vs. various CSMs, especially since people type passwords into one of those screens

**Comment by CowboyLva** [comment]
> > 5. More shadow angle customization(as of now sun is only Northwards, even at 0 the shade side is South).
> 
> Sounds like a bug to me?
Values allowed between 0 and 60
0 - Sun above
60 - Sun at a low elevation angle, north of side of sky at noon
It's not a bug it's a bad feature. I want to set the Sun to shine from South, like it normally does on the Northern Hemisphere.

Pay attention to time and where shadow is

0 degrees
![screenshot_20221022_145006](https://user-images.githubusercontent.com/108756200/197337419-804e8087-5073-4610-9aee-e320c0bfdc22.png)
![screenshot_20221022_145013](https://user-images.githubusercontent.com/108756200/197337426-04f462ce-c6cf-4c50-a247-a1a3487a9b7e.png)

60 degrees
![screenshot_20221022_145200](https://user-images.githubusercontent.com/108756200/197337482-ef9fb35e-ac04-430d-ab45-b9224d12efeb.png)

What is needed is the possibility for the axial tilt to be in the negatives. Since the reason Sun goes through the north is because the angle is allowed only positive numbers (+Z axis).

P.S.
Version ran is minetest 5.6.1 stable.

**Comment by x2048** [comment]
@CowboyLva Can you register an issue for this? It can be fixed sooner than 5.8.

**Comment by CowboyLva** [comment]
Did so just now. Kinda strange since it seemed so obvious, expected already for someone to submit one.
Anyway thanks.

**Comment by Desour** [comment]
Could please someone explain to me how these brainstorm discussions contribute to the final roadmap?
@Zughy's table makes it look like it was some sort of democratic decision, but that makes no sense. Coredevs decide on the roadmap (obviously), so how many people commented on some thing is quite irrelevant, especially if the comment came from someone who hasn't ever and most likely will not in the near future do any contribution, and who hence won't implement their requested features themselves.

**Comment by appgurueu** [comment]
> Could please someone explain to me how these brainstorm discussions contribute to the final roadmap? @Zughy's table makes it look like it was some sort of democratic decision, but that makes no sense. Coredevs decide on the roadmap (obviously), so how many people commented on some thing is quite irrelevant, especially if the comment came from someone who hasn't ever and most likely will not in the near future do any contribution, and who hence won't implement their requested features themselves.

Assuming some core devs being indifferent to what they want to implement next (or caring about what modders want next), this brainstorm can give them a hint on what modders would like to see implemented.

**Comment by snowyu** [comment]
#### Different dimensions

> @nerzhul: Multiple maps, with different mapgen for example (breaking, obvisouly, if server has multiple layers

> @CowboyLva:  Better map related stuff - Optimized mapgen, definable world size, larger possible worlds at least xz axes, per map configuration menu(e.g. how far each axis will generate), easily customizable generation distance(not render distance).

Scalable Global Coordinate and Local Coordinate. Teleport from server to another...

#### Better Enties

> @appgurueu:  Entity Performance Improvements

> @nerzhul: better enties, with more intelligent API to transfer actions from server to client. We should not have for example to run per movement code server side and set each movement to client, a jump should be a jump, a move from A to B if using the path finder should use path finder on client. Obviously with entity pos resync every 5 sec or something. This world is so boring because we don't have more high level API :) look at all those minecraft things (breaking, obviously, if a server uses it)

The ECS(Entity-Component-System Architectural Pattern) may be a good start:

* [Flecs](https://github.com/SanderMertens/flecs) MIT C99/CPP11
* [EnTT](https://github.com/skypjack/entt) MIT CPP17
* [ECS-LUA](https://github.com/nidorx/ecs-lua) MIT LUA

Better enties can use Grouped Nodes With Inherits to make player, NPC, trees and tools etc objects. And can apply skeletal animations, physical engine to them.

![grouped-node](https://user-images.githubusercontent.com/327887/197452937-8ca1e784-ea4c-462f-990d-7b8f3d77b94d.gif)


eg:

* [ozz-animation](https://guillaumeblanc.github.io/ozz-animation/samples/attach/) MIT CPP
* [Jolt Physics](https://github.com/jrouwe/JoltPhysics)
* [ReactPhysics3D](https://github.com/DanielChappuis/reactphysics3d?)
* [Newton Dy
![output1](https://user-images.githubusercontent.com/327887/197452889-787d7d18-2434-4c73-875b-b8d61b35a3be.gif)
namics](https://github.com/MADEAPPS/newton-dynamics)
* [PhysX](https://github.com/NVIDIAGameWorks/PhysX)

The texture could be image, animation or video...

Make minetest a true voxel game engine library first.

**Comment by immagiov4** [comment]
**Improve UI**
The graphical interface is the first thing a player sees when opening minetest and, as we know and have discussed multiple times, Minetest's default UI (and UX, often) is... bad. 

**SSCSM**
Without this MT can't be a proper engine imo.

**Different dimensions**
I agree with Zughy, disposable worlds would be so handy for things like minigames.

**Comment by v-rob** [comment]
#### Formspec/HUD replacement

This is an "of course" coming from me, but it is my basic priority in Minetest.  It isn't stagnant, however, as shown by #12926, so it's not like I'm suggesting it without planning to work on it.

#### Dehardcoding

I agree with sfan5--Minetest has far too many hardcoded features, and to support more general games, we should make it possible for them to disable the assumptions that Minetest currently makes.

#### Entities

Because they're garbage.  Performance, of course, since entities are heavy.  We need at least some better ways of moving them than just raw `set_pos`, `set_velocity`, and `set_acceleration`.  These are only useful for very simple use-cases; higher-level motion with them from the server is janky.  I think we need more specialized functions for interpolated motion of entities along paths.

**Comment by oong819** [comment]
add content creator features

Replay mod like

**Comment by oong819** [comment]
# Video Render From World Data
like Replay Mod in Minecraft, this will be useful for content creator

# In-app Text Editor for Mod Creator
when making mods, i hate constantly switching app to edit lua file and testing in game

# Accessibility Features
Left hand controls, subtitle, etc.

# More Focus on MTG
i know MTG have been switched to "dead" mode but Minetest dont have enough developer to fill the gap. MTG are stable and light-weight but no longer receive major update really make me sad

# Reduce World File Size
idk why but Minetest world file are a lot bigger than similar game

# Better Touchscreen Support
idk what to say but use a phone to play pvp and u will know

# End
i already have a commeny here but i have some new ideas

Minetest C++ Edition and Minetest Mese Edition XD

**Comment by appgurueu** [comment]
> ###### Video Render From World Data
> like Replay Mod in Minecraft, this will be useful for content creator

Pretty infeasible. If anything, Minetest should give mods the tools to implement this themselves.

> ###### In-app Text Editor for Mod Creator
> when making mods, i hate constantly switching app to edit lua file and testing in game

100% infeasible and not worth the effort in the slightest. Formspecs are a mess incapable of implementing text editing with syntax highlighting. And even if it were possible, such a text editor would always be way behind proper editors like VSC or Vim. "Hot reload" also is nontrivial to implement.

Get a proper window manager. This isn't Boblox (I've heard that Boblox Studio or whatever it's called also sucks BTW). We're not here to reinvent editors.

> ###### Accessibility Features
> Left hand controls, subtitle, etc.

Valid point.

> ###### More Focus on MTG
> 
> i know MTG have been switched to "dead" mode but Minetest dont have enough developer to fill the gap. MTG are stable and light-weight but no longer receive major update really make me sad

MTG maintainer here. We also don't have the developers to maintain MTG. It may be stable, but it certainly is not lightweight (bloated codebase of ~30kloc which needs a massive refactoring basically).

That said, if MTG ever overcomes its technical debt (highly unlikely that this happens anytime soon if ever though), it *might* re-enter active development (after maintainer discussion).

Ideally fresh games will take the place of MTG.

> ###### Reduce World File Size
> idk why but Minetest world file are a lot bigger than similar game

Be more precise: What do you consider a "Minetest world file"? Worldsize varies greatly depending on mapgen and mods (nodes + eventually metadata). I don't think worlds are particularly large (and I don't really see how we could compress them much further).

> ###### Better Touchscreen Support
> idk what to say but use a phone to play pvp and u will know

Agreed.

> ###### End
> i already have a commeny here but i have some new ideas  
> Minetest C++ Edition and Minetest Mese Edition XD

Sorry what? Minetest *is* already written in C++ (plus some Lua). Getting rid of the Lua part in favor of C++ would earn you... nothing? Except for probably "performance" in exchange for a major effort and incompatibility with mods. And would what the "Mese Edition" even be?

**Comment by oong819** [comment]
> > ###### Video Render From World Data
> > like Replay Mod in Minecraft, this will be useful for content creator
> 
> Pretty infeasible. If anything, Minetest should give mods the tools to implement this themselves.
> 
> > ###### In-app Text Editor for Mod Creator
> > when making mods, i hate constantly switching app to edit lua file and testing in game
> 
> 100% infeasible and not worth the effort in the slightest. Formspecs are a mess incapable of implementing text editing with syntax highlighting. And even if it were possible, such a text editor would always be way behind proper editors like VSC or Vim. "Hot reload" also is nontrivial to implement.
> 
> Get a proper window manager. This isn't Boblox (I've heard that Boblox Studio or whatever it's called also sucks BTW). We're not here to reinvent editors.
> 
> > ###### Accessibility Features
> > Left hand controls, subtitle, etc.
> 
> Valid point.
> 
> > ###### More Focus on MTG
> > i know MTG have been switched to "dead" mode but Minetest dont have enough developer to fill the gap. MTG are stable and light-weight but no longer receive major update really make me sad
> 
> MTG maintainer here. We also don't have the developers to maintain MTG. It may be stable, but it certainly is not lightweight (bloated codebase of ~30kloc which needs a massive refactoring basically).
> 
> That said, if MTG ever overcomes its technical debt (highly unlikely that this happens anytime soon if ever though), it _might_ re-enter active development (after maintainer discussion).
> 
> Ideally fresh games will take the place of MTG.
> 
> > ###### Reduce World File Size
> > idk why but Minetest world file are a lot bigger than similar game
> 
> Be more precise: What do you consider a "Minetest world file"? Worldsize varies greatly depending on mapgen and mods (nodes + eventually metadata). I don't think worlds are particularly large (and I don't really see how we could compress them much further).
> 
> > ###### Better Touchscreen Support
> > idk what to say but use a phone to play pvp and u will know
> 
> Agreed.
> 
> > ###### End
> > i already have a commeny here but i have some new ideas
> > Minetest C++ Edition and Minetest Mese Edition XD
> 
> Sorry what? Minetest _is_ already written in C++ (plus some Lua). Getting rid of the Lua part in favor of C++ would earn you... nothing? Except for probably "performance" in exchange for a major effort and incompatibility with mods. And would what the "Mese Edition" even be?

# ![Screenshot_20221114-202632](https://user-images.githubusercontent.com/76146430/201672076-f9aeccf5-2212-4b9d-9f0e-e40d3ccb5b23.png)


# 2
about the "replay" i think converting block data to a 3d mesh in obj for 3th party app will be easy?
#
and the last thing is a joke sorry for the confusion

**Comment by cranezhou** [comment]
In a short summary, I want to emphasize some points from my perspective view. (only personal suggestions)
1) minetest game engine shall provide a better technical solution(s) and support (documents) for the mod developers on different os platform. This may help minetest software proudct  to compete with minecraft and other similar products. (generally and legally)
2) the product (project) manager of minetest shall align the requirements with the current limited development resource on each milestone versions. (v5.8+)
3) If the minetest software product could benefit the traditional education (online and offline), that will be great. 
(not only the community)

Thank you very much for your help and support.

Best Regards,

Crane Zhou (2022.11.15 /  Tuesday morning）

**Comment by cranezhou** [comment]
I am involved in this discusstion slightly.

I will be silent if there is a voting process.

**Comment by MisterE123** [comment]
> I am involved in this discussion slightly.
> 
> I will be silent if there is a voting process.

it was a figure of speech, not a real vote. The devs want to here what modders want.

**Comment by fluxionary** [comment]
> # In-app Text Editor for Mod Creator
> 
> when making mods, i hate constantly switching app to edit lua file and testing in game

i feel your pain

> such a text editor would always be way behind proper editors like VSC or Vim.

such a solution isn't fully realizable w/ the current APIs, but you *can* do quite a bit in-game. there are currently ways to live-patch servers, though they're limited

* https://forum.minetest.net/viewtopic.php?t=20161
* https://content.minetest.net/packages/Just_Visiting/formspec_editor/
* https://content.minetest.net/packages/luk3yx/snippets/

in particular, i use snippets to monkey patch live servers almost constantly, but it's got a lot of room for improvement, and isn't oriented to new contributors. 5

> Formspecs are a mess incapable of implementing text editing with syntax highlighting. 

we do need a replacement for formspec, for many reasons, this might as well be another one of them. 

> And even if it were possible, such a text editor would always be way behind proper editors like VSC or Vim. 

that doesn't mean that we couldn't make a way to create a basic editor that properly did things like copy/paste, undo, syntax checking, spell checking, etc. etc.. 

> We're not here to reinvent editors.

no, but we're making (or you're making and i'm recently contributing to) an open-source sandbox game which derives most of its content from third party mods. it'd be really awesome if we could close that loop by allowing people to create mods in-game and test how things changed when they re-defined various stuff. you're absolutely right that that's beyond project scope for the engine at its current level of development, but it'd be awesome if it existed, and it's certainly not impossible. i see nothing wrong w/ putting dreams on the want-list.

**Comment by appgurueu** [comment]
> no, but we're making (or you're making and i'm recently contributing to) an open-source sandbox game which derives most of its content from third party mods. it'd be really awesome if we could close that loop by allowing people to create mods in-game and test how things changed when they re-defined various stuff. you're absolutely right that that's beyond project scope for the engine at its current level of development, but it'd be awesome if it existed, and it's certainly not impossible. i see nothing wrong w/ putting dreams on the want-list.

Yes, but this is the wrong dream to have. The real dream should be hot-reloading (e.g. by watching files), not an editor inside Minetest. Consider Flutter: Imagine instead of proper hot-reloading triggered from your editor you had an in-app editor. How awful would that be?

**Comment by velartrill** [comment]
the text-editing primitives do need a *serious* overhaul (at the absolute least, not resetting their contents whenever you resize the game window), but only for the sake of books, signs, and other in-world messages. that said, trying to cram a lua IDE into minetest (whether as a core feature or a mod) is a bizarre idea that runs entirely counter to the principle of "do one thing and do it well." minetest is a game *engine* like love2d, not a wysiwyg RAD studio like Godot or RPG Maker. completely different class of application with fundamentally different needs -- game engines supply their own bespoke UI kits, take over the desktop to monopolize the user's attention, and have high CPU usage due to redrawing unconditionally every 60th of a second or so; studios need to play nice with the desktop environment & other applications, and *block* for input or damage events so they don't hog the CPU and GPU with pointless render calls, and so they can react quickly and smoothly. there's also the whole looming spectre of security & authentication; the idea of adding a function that allows admins to directly edit sensitive files on a remote server with the daemon's permissions from within the game fills me with dread.

a more realistic and useful feature (which could be added right away, unlike hot reloading which is distinctly nontrivial due to minetest's… aggressively stateful API architecture) might be a function key to quick-reload the current game, so you don't have to open the menu, click quit, wait for the main menu to open, then select the world and click play/log back in to a server. it could even send a restart signal to remote servers if the user is connected with the right privileges, and maybe add an option to cache the user's password in memory so he doesn't have to re-enter it when the remote server comes back up. *that* is something that would save me a lot of time and exasperation working on the mod i maintain. if it could reload the game without dumping & reloading the world into memory, that would be even better.

**Comment by MathPlusEnglish** [comment]
I think adding features to Minetest Game would be nice.

**Comment by runsy** [comment]
My ideas:

1) Implementing a GUI change is easy and straightforward. The easiest. I don't know why it has been done already, really.
I'm preparing a soon to be released prototype GUI design for Minetets, but coolZ and cuteZ.

2) Multidimensions mod could be implemented in the engine, it's already written, it's just pass it to C++, a max of 3 layers, normal world, underworld (nether) and aether. With an indestructible node layer for nop to be able to access, select width of this layer if such. Simple, isn't it? Well, I don't know why it hasn't been implemented yet.

3) SSCMS I don't know what it is and neither do 99% of the people, I have no opinion.

4) Performance, through the improvement of things already done, is ultimately being done.

5) The mobs entities have been traditionally neglected by the developers, a little more love would not hurt.

**Comment by ezileicorreia** [comment]
Hello survivors of reality, who still retain the ability to dream.

### Support for Mumble (Positional Voice Chat)

Mumble is an open source voice chat software intended primarily for use during gameplay. It's been discussed before, but good ideas don't always pass.

### Colored Lighting

For a full game engine colored lighting is key, I believe there is a way to add it to Minetest without sacrificing performance, perhaps through an API.

### Animation for Sneak

It took me a long time to discover that Minetest has this feature, but it lacks animation, so I would like to see the cover animation already in the next versions.

**Comment by fluxionary** [comment]
> ### Animation for Sneak
> 
> It took me a long time to discover that Minetest has this feature, but it lacks animation, so I would like to see the cover animation already in the next versions.

this has been done w/ a mod, though i haven't tested it myself: https://content.minetest.net/packages/niwla23/visible_sneak/

note that mods that alter the player animation and player model often don't work well together.

**Comment by SmallJoker** [comment]
https://irc.minetest.net/minetest-dev/2022-12-04#i_6038232

> \<rubenwardy\> 	discussing a new roadmap before discussing how the current one went is a bad idea imo
> [...]
> \<Krock\> 	work out why? because people implement what they prefer

I can see many wishes towards a better UI and SSCSM for more modding capabilities - but who is willing to actually work on these points? For example, the main menu is on a good path to improve thanks to rubenwardy's pending PR and the rendering part has seen major improvements thanks to x2048.

Rather than posting a wish list, I believe it would be better to ask what people are interested to implement or fix. Developers often have a code area they take care of, which also worked and still works pretty well. An example from the core team: (including past contributors)

 * paramat: mapgen
 * sfan5, myself: modding features/fixes, maintenance
 * TurkeyMcMac: modding features/fixes
 * nerzhul: networking, Android
 * rubenwardy: UI, CDB, Android
 * x2048: shaders, rendering

Even though nobody takes care of SSCSM right now, you might still find answers to implementation questions in the development IRC channel. As for the roadmap - I propose to update it according to the progress so far, and to include SSCSM in hopes that it will be addressed in the next few years.

**Comment by runsy** [comment]
> ather than posting a wish list, I believe it would be better to ask what people are interested to implement or fix. Developers often have a code area they take care of, which also worked and still works pretty well. An example from the core team: (including past contributors)

There you have hit it. I didn't mean it, for the umpteenth time and more people.

The shading was done by the enormous work of x248, without him there would be no shadows for at least ten years. And for his effort, since at the beginning nobody supported him, in fact, everything was a hindrance.

The problem with brainstorming and roadmaps and other crap is that they are empty words. In the end for nothing.

For example I would like to see it supported:
https://github.com/minetest/minetest/pull/12928
https://github.com/minetest/minetest/pull/11016

The gui has been delayed for 2 years, because no one likes anything and calls the other tacky. There is fear, a lot of fear to implement or change the GUI.

**Comment by velartrill** [comment]
i'm happy to keep iterating on the particle system, if people have requests for improvements. >2-keyframe animations are the only thing on my to-do list currently

**Comment by tigercoding56** [comment]
someone on the[ forums](https://forum.minetest.net/viewtopic.php?p=418797#p418797)  asked me to add this :
asynchronous re-projection - basically another technique to make the game/engine appear smoother when using low-end hardware

**Comment by rubenwardy** [comment]
This is not a thread to make random feature suggestions, please use the issue tracker for that

**Comment by ezileicorreia** [comment]
## Support for Mumble (Positional Voice Chat)

Mumble is an open source voice chat software intended primarily for use during gameplay.

## Colorful Lighting

For a full game engine colored lighting is key, I believe there is a way to add it to Minetest without sacrificing performance.

## Support for Google VR

A virtual reality experience that starts with a simple screen that anyone can build or buy.

**Comment by ezileicorreia** [comment]
### Support for Mumble (Positional Voice Chat)

Mumble is an open source voice chat software intended primarily for use during gameplay.

### Colorful Lighting

For a full game engine colored lighting is key, I believe there is a way to add it to Minetest without sacrificing performance.

### Support for Google VR

A virtual reality experience that starts with a simple screen that anyone can build or buy.

**Comment by jingkaimori** [comment]
## rework data sending mechanism to send arbitrary dictionary-like data.
new data sending mechanism should behave like RPC(remote process calling) protocol, such as msgpack, grpc or protobuf. These RPC protocols will generate suitable serializing process for cpp classes and structs, instead of creating different stringtify method for every transferable class. With auto-fitted serializing process, it will be convenient to modify data structs transferring between client and server, and reduce usage of "service" and "handler" introduced by adding new field.

## rework lua glues to transform known data-structures between scripting api and engine.
Similar as above issue, every type shared between lua script and cpp engine is wrapped manually. When modifying these types, we should modify existing glues which transform cpp objects into lua objects. Minetest can try to map arbitrary cpp object to lua object depends on its public members.

## Use TypedLua to define apis
It's inconvenient for mod developers to invoke a minetest api. The only document about the api is a huge text file named as  `api.txt`. When calling api function, developers may be confused by param type and return type of an function, and told wrong usage only after loading in minetest game. Lua type checker such as [Teal](https://github.com/teal-language/tl) or [TypedLua](https://github.com/andremm/typedlua) can reduce pains guessing behavior of minetest lua api.

**Comment by AFCMS** [comment]
> ## Use TypedLua to define apis
> 
> It's inconvenient for mod developers to invoke a minetest api. The only document about the api is a huge text file named as `api.txt`. When calling api function, developers may be confused by param type and return type of an function, and told wrong usage only after loading in minetest game. Lua type checker such as [Teal](https://github.com/teal-language/tl) or [TypedLua](https://github.com/andremm/typedlua) can reduce pains guessing behavior of minetest lua api.
> 
> [Existing work on minetest typedef using teal](https://github.com/teal-language/teal-types/pull/5).

While I agree some better doc system should be added, I don't think an official system should be added.

I would rather link in website, wiki, etc to community made API headers.

There are multiple other solutions for API typing in Lua, depending of the devs preferences like [EmmyLua](https://github.com/EmmyLua) and [derivated](https://github.com/sumneko/lua-language-server)
([Existing work for Minetest](https://github.com/AFCMS/minetest_vscode_lua/tree/dev))

**Comment by oong819** [comment]
### Law, policy and license
Headache first then fun, this affect all Minetest player and modder. To keep everyone happy, we have to be clear about those things
### Make modder comfort
Minetest is a game engine and how its look, how its feel heavily depend on the ContentDB, however if all the mod and game developer leave, player will be leave alone with MTG (this will not likely happen but its will happen in some smaller scale if something went wrong)
### Improve UI
#### improve UI on touchscreen
Minetest Android setting tab is a pain, 
 espeacially the scroll text
#### more flexible to screen size
some text formspec got cut off both in some mod and main menu
### Dehardcoding
everyone love this, help make Minetest easier to edit!
### Anti cheat
i love how Minetest prevent Xray, but i can still see people use fly hack with modded client, should have some options to prevent this since Minetest was an open source game
### Encrypt data before being sent
Someone can stalk on weak network or even edit data!
<hr>

thats my thoughts, feel free to debate and change my or point out if i wrong, after TenPlus1 leave ContentDB, i was shocked on how easy things can go wrong and the playerbase got affected but i see he on ContentDB now but who know things like that can happen in future

Also changed my mind about MTG its should be like that sry for being annoying

**Comment by appgurueu** [comment]
> [6 points], in addition to [5 points] posted earlier

Please re-read the post carefully:

> please list **3 things** that you think should be prioritised in the medium-long term

This is about **priorities**, not dumping everything that comes to mind. Especially your "law, policy and license" point looks like a joke to me.

**Comment by Zughy** [comment]
Hello, since rubenwardy had already warned about how this thread works and yet someone kept posting things not in line with the thread, I'm sorry to announce I'll delete without any further notice any message that keeps missing the point. Yes, you can discuss what people suggest (possibly in 1-2 comments top), just, this is not the place to push for single specific features or to suggest 6 different things at once. Instructions are easy, c'mon people, I believe in you!

**Comment by Zughy** [comment]
> <rubenwardy> discussing a new roadmap before discussing how the current one went is a bad idea imo

You could have stopped me when I asked on IRC if somebody was against this - and no one said anything. I'll take it in consideration for the next roadmap.

> I can see many wishes towards a better UI and SSCSM for more modding capabilities - but who is willing to actually work on these points? 

The point of the thread is not imposing to core devs what they should do, but rather telling them what the community (at least the one having a GH account) would like to see. It's not that different from [this](https://forum.minetest.net/viewtopic.php?f=3&t=24742). It's also a mental exercise for core devs to help them understand better what are their single priorities - hence my pushing for a comment from each one of you - let alone the discussion that's generated between single contributors. Then, in the end, it's of course up to core devs decide where to focus on, but at least there is a dialogue between the parties.

> Rather than posting a wish list, I believe it would be better to ask what people are interested to implement or fix

This might work on paper, but we've already seen core devs joining the team and then disappearing after a short while. What I mean is, people could say a lot of things, potentially creating expectations, but in the end just a bunch of them would probably end up opening a PR

**Comment by Anon7250** [comment]
I have a random idea I've always wanted for both Minetest and Minecraft, (but it suits Minetest better):

Server discovery is hard for new players. Minetest has the "Join Game" tab. It's better than Minecraft (you'd have to visit random websites where people/bots "vote" for the top servers).

But we can improve the "join game" tab if the game can take 2D or even 3D "screenshots" of the best parts of each server, and **if players can see these screenshots without having to register for an account**.

Perhaps we can even make an official "Find a World" server where each chunk features the best content from a different server. Server owners would choose a chunk to represent their server. And the game can take 3D screenshots of this chunk every day, and update the corresponding chunk on "Find a World" server.

Just an idea. I don't know how / whether it will be easy to implement.

**Comment by oong819** [comment]
i think you should open a feature request for that

(Off-Topic)

**Comment by tigercoding56** [comment]
i do not see the issue with discussing ideas for things that could be added to the roadmap on the roadmap brainstorm (definition : to come up with ideas ) discussion so rather than wasting time here , maybe you have some other issues you may want to spend time on to do this look at the "issues" tab

Sent with [quark](https://mail.proton.me)[Mail](https://proton.me/) unsecure email.

------- Original Message -------

**Comment by tigercoding56** [comment]
ok ,
i hate to inform you
but i followed the instructions
someone on the forums gave to
post it here so yeah
also i am pretty sure i did not
post this suggestion on 6 different places

Sent with [Photon Mail](https://proton.me/) unsecured email.

------- Original Message -------

**Comment by tigercoding56** [comment]
well i do not mean to be intrusive and rather than
adding it to the already large pile of issues i want
to add it to the maybe do later if time pile

Sent with [electron Mail](https://proton.me/) unsafe email.

------- Original Message -------

**Comment by runsy** [comment]
3 things I want now for 5.7:

- Dual wielding.
- Texture node changing in real time.
- 3D Line rendering.

**Comment by michieal** [comment]
A really good addition, in my humble opinion, would be to add in a dual-wield mechanic to minetest, and grant api access to it. I know that Mineclone 2 would love to have that, as we are doing a hack to add it in.

**Comment by michieal** [comment]
> That's already planned to get added for 5.7.

I didn't know that. Thanks!

Then, how about the Revamped UI for Minetest? That would also be a great thing to have.

**Comment by michieal** [comment]
My first Suggestion is the Revamped UI for Minetest.
My second suggestion is Tips during the loading screen.

**Comment by michieal** [comment]
Third Suggestion is Support for larger structures being spawned. That way, we can make something larger than a small  building.

**Comment by fluxionary** [comment]
> Third Suggestion is Support for larger structures being spawned. That way, we can make something larger than a small building.

that can already be done via schematics. how large were you thinking? you can manipulate dozens of mapblocks w/ worldedit w/out much issue, and hundreds if you're willing to wait a bit. 

or were you thinking large procedurally generated structures? that's also quite possible already, though i don't think anyone's made a framework for generating such things easily. there's [existing opensource code for minecraft](https://github.com/TeamTwilight/twilightforest/tree/1.19.x/src/main/java/twilightforest/world/components/structures) that someone could use for inspiration if they want to design such a mod.

**Comment by nephele-gh** [comment]
> As much as burning formspecs to the ground is a noble goal, this is not that simple. Tons of games and mods depend on this, so a complete removal means breaking almost all games and mods ever made. Nothing short of a hard fork will get you there. A full removal of formspecs is not going to happen.

An api for drawing to the screen like the [love2d api](https://love2d.org/wiki/love.graphics) Would be enough, then a formspec compatible implementation can be build ontop of this.
Alternatively this can simply like sidebyside with the current api. FOr my part I have my own lua UI toolkit i'd love to use for minetest but it really requires a more direct api like the one of garrys mod or the aformentioned love2d one.

The priorities I'd like for my gamemode to work better are mainly stuff that has been talked about above.

* Dehardcoding engine behaviour (like adding random inventory lists to the player, playing the red damage animation even if falldamage is 0 etc.) 

* A direct api to draw to the HUD and to implement inventories etc. like mentioned above 

* A way for gamemodes to more directly influence game settings in the scope of the game (Like item time to live) without affecting other games

(also different dimensions would be awesome to implement, I'd definetely make use of them. but you asked for three items so :)

**Comment by velartrill** [comment]
@nephele-gh love2d these days is sort of a weird awkward wrapper around opengl. it *would* be pretty neat if client-side scripts could do their own gl-level rendering, tho i don't know if there's any way to sandbox that so it doesn't interfere with the GL state of the client itself, and that's not even getting into the whole issue with the multiple GL profiles that minetest has to support

come to think of it most of the discussion ive seen so far around the formspec replacement has been from the perspective of a new UI toolkit. i haven't seen any talk about a general-purpose canvas mechanism, which it seems like a game engine should have after some sort of fashion, so it might be worth creating a new issue for this -- it's the kind of thing that's going to take a lot of discussion to work out, particularly since there's already a formspec replacement underway at #12926 and unfortunately it's based on directly decoding and rendering a JSON tree (personally i would prefer the kind of approach you suggest but i suspect that ship has sailed. getting some kind of canvas element in might be doable tho)

**Comment by v-rob** [comment]
If you can find a way to do direct drawing without client-side modding, I'd love to hear it.  I'd love the idea of direct drawing.  However, I strongly suspect it's impossible.  Otherwise, SSCSM will need to be figured out first.  However, this is offtopic for this thread.

**Comment by MathPlusEnglish** [comment]
Brainstorm ideas:

Make it impossible to ban evade in servers by default.
Be able to code in-game. 
Be able to turn players into camera/viewer with interact. Example: https://youtu.be/6UTXzjcqhS8.

**Comment by fluxionary** [comment]
> Make it impossible to ban evade in servers by default. 

there is no general programmatic solution to ban evasion. doing so would require a fully reliable identification system, which is far outside the scope of minetest, and also not in the spirit of free software. [i've done some work](https://github.com/fluxionary/minetest-verbana) on creating a powerful verification/banning mod, but currently it only works w/ IPv4, so it doesn't work w/ a lot of modern servers. 

> Be able to code in-game. 

the [mesecons luacontroller](https://mesecons.net/luacontroller/) is awesome, and there's plenty of didactic mods w/ a codeing bent (e.g. [vbots](https://content.minetest.net/packages/Nigel/vbots/))

> A way for gamemodes to more directly influence game settings in the scope of the game (Like item time to live) without affecting other games

it's currently possible to override settings w/out affecting other worlds, but most mods only pay attention to global settings. it's possible to create APIs to allow modification of setting parameters which mods can respond to, but there's no standard for doing that. [i created an API for mesecons_debug](https://github.com/minetest-monitoring/mesecons_debug/blob/d0787b0a9bc55aaccfe5400a710ea7b531926ac7/settings.lua#L53-L66) to allow modifying game settings, but more work needs to be done to make this a functional standard API . we need an RFC system for minetest...

> Be able to turn players into camera/viewer with interact. Example: https://youtu.be/6UTXzjcqhS8.

i'm pretty sure some work is being done on this, but i'm personally failing to bring up the links to the relevant issues, sorry.

**Comment by EntityinArray** [comment]
my 5 cents - searching mods on content.minetest.net via mod id, ability to see and instantly install optional mod dependencies

**Comment by EntityinArray** [comment]
I have another suggestion regarding search.
I think it would be better to have 2 separate fields:
- Mod dependencies
- Game dependencies

The latter one will allow modders to declare which games this mod is designed for. It will also allow users to search mods for a specific game.

**Comment by EntityinArray** [comment]
> @EntityinArray already exists: https://content.minetest.net/help/game_support/
> Example of game hub: https://content.minetest.net/packages/Warr1024/nodecore/hub/

Very cool! Then I think we should integrate supported_games filter into Browse Online Content menu

**Comment by lhofhansl** [comment]
+1 on entity improvement:
- better attachment, without jitter or rubberbanding (two separate issues)
- performance, especially collision detection with many entities
- proper player_pos when attached
- better camera control
- entity path apis 
- etc

Also
- making sense of the various distance settings. It's very confusing to a new user.
- remove unneeded config config options, calculate them instead
- multi-threaded Lua map gen
- reflective water!
- godrays

**Comment by Zughy** [comment]
Here ya, here ya!

After having talked about the roadmap in a couple of meetings, we've basically agreed to keep the current the roadmap as-is:
* Rendering/Graphics improvements (2 votes)
* Internal code refactoring (4 votes)
* UI Improvements (8 votes)
* Object and entity improvements (5 votes)

About the point with more votes (SSCSM, 9 votes), there seems to be the will to work on it (https://github.com/minetest/minetest/pull/13046#issuecomment-1510455114), but at the moment it's not considered as urgent as the rest.

Thank you everyone for your feedback and see you in a couple of releases!
